packages/iov-bns/src/bnsconnection.ts

```diff
@@ -190,55 +193,68 @@ export class BnsConnection implements BcpAtomicSwapConnection {
-            next: async blockHeader => {
-              const event: BcpBlockInfo = {
-                state: BcpTransactionState.InBlock,
+    const firstEvent: BlockInfo = { state: TransactionState.Pending };
+    let lastEventSent: BlockInfo = firstEvent;
+    const blockInfoProducer = new DefaultValueProducer<BlockInfo>(firstEvent, {
+      onStarted: () => {
+        transactionSubscription = this.liveTx({ id: transactionId }).subscribe({
+          next: searchResult => {
+            if (isFailedTransaction(searchResult)) {
+              const errorEvent: BlockInfoFailed = {
+                state: TransactionState.Failed,
+                code: searchResult.code,
+                log: searchResult.log,
+              };
+              blockInfoProducer.update(errorEvent);
+              lastEventSent = errorEvent;
+              return;
+            }
+
+            // Don't do any heavy work (like subscribing to block headers) before we got the
+            // search result.
+
+            const transactionHeight = searchResult.height;
+            const transactionResult = searchResult.result;
+
+            {
+              const inBlockEvent: BlockInfoSucceeded = {
+                state: TransactionState.Succeeded,
                 height: transactionHeight,
-                confirmations: blockHeader.height - transactionHeight + 1,
+                confirmations: 1,
                 result: transactionResult,
               };
-
-              if (!equal(event, lastEventSent)) {
-                blockInfoProducer.update(event);
-                lastEventSent = event;
-              }
-            },
-            complete: () => blockInfoProducer.error("Block header stream stopped. This must not happen."),
-            error: error => blockInfoProducer.error(error),
-          });
-        } catch (error) {
-          blockInfoProducer.error(error);
-        }
+              blockInfoProducer.update(inBlockEvent);
+              lastEventSent = inBlockEvent;
+            }
+
+            blockHeadersSubscription = this.watchBlockHeaders().subscribe({
+              next: async blockHeader => {
+                const event: BlockInfo = {
+                  state: TransactionState.Succeeded,
+                  height: transactionHeight,
+                  confirmations: blockHeader.height - transactionHeight + 1,
+                  result: transactionResult,
+                };
+
+                if (!equal(event, lastEventSent)) {
+                  blockInfoProducer.update(event);
+                  lastEventSent = event;
+                }
+              },
+              complete: () => blockInfoProducer.error("Block header stream stopped. This must not happen."),
+              error: error => blockInfoProducer.error(error),
+            });
+          },
+          error: error => blockInfoProducer.error(error),
+        });
       },
       onStop: () => {
+        if (transactionSubscription) {
```

Comment:
I guess we rely on the client unsubscribing after a failure is returned for the proper processing as I mentioned above. I guess this works, seems a bit harder to follow logic, as it depends on the actions of other parts.

Reply:
Nope, the client does not need to unsubscribe. In case of FailedTransaction, there is one event, an open stream but no further events.

We don't waste resources in this case because `blockHeadersSubscription = this.watchBlockHeaders().subscribe({` is never started for FailedTransaction.